onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -125,8 +126,71 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
                                    std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
-                                   const logging::Logger& logger) {
+                                   const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                   const logging::Logger& logger,
+                                   int64_t& run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  if (run_id == DEFAULT_RUN_ID) {
+    ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+    return Execute(session_state, feeds, fetch_mlvalue_idxs, fetches, logger, frame, 0, exec_plan_size);
+  } else {
+    // Partial graph execution frame management code.
+    ExecutionFrame* frame;
+    std::map<int64_t, std::unique_ptr<onnxruntime::ExecutionFrame>>::iterator it;
```

Comment:
It is used in the end after both the branches to erase the run after full graph has been executed, this is to prevent additional look up.

---
In reply to: [598994948](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r598994948) [](ancestors = 598994948)

Reply:
It looks like this code is gone now, but there *was* a data race in the old code.

You can't call `find()` under the mutex, release it, and then later call `erase()` on the same iterator (even if you re-acquire the mutex) because another thread might have mutated the map and invalidated the iterator.

Reply:
That is true, hence I centralized all of the atomic operations under PartialGraphExecutionManager.

---
In reply to: [600946197](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r600946197) [](ancestors = 600946197)